atomics/include/desul/atomics/Fetch_Op_HIP.hpp

```diff
@@ -9,99 +9,108 @@ SPDX-License-Identifier: (BSD-3-Clause)
+  }                                                                               \
+  template <class MemoryOrder, class MemoryScope>                                 \
+  __device__ inline T device_atomic_fetch_dec(T* ptr, MemoryOrder, MemoryScope) { \
+    return __hip_atomic_fetch_add(ptr,                                            \
+                                  -1,                                             \
+                                  HIPMemoryOrder<MemoryOrder>::value,             \
+                                  HIPMemoryScope<MemoryScope>::value);            \
+  }
 
-inline __device__                int device_atomic_fetch_dec(               int* ptr,                         MemoryOrderRelaxed, MemoryScopeDevice) { return atomicSub(ptr,  1  ); }
-inline __device__       unsigned int device_atomic_fetch_dec(      unsigned int* ptr,                         MemoryOrderRelaxed, MemoryScopeDevice) { return atomicSub(ptr,  1u ); }
-inline __device__ unsigned long long device_atomic_fetch_dec(unsigned long long* ptr,                         MemoryOrderRelaxed, MemoryScopeDevice) { return atomicAdd(ptr, -1  ); }
+DESUL_IMPL_HIP_ATOMIC_FETCH_INC(int)
+DESUL_IMPL_HIP_ATOMIC_FETCH_INC(long long)
+DESUL_IMPL_HIP_ATOMIC_FETCH_INC(unsigned int)
+DESUL_IMPL_HIP_ATOMIC_FETCH_INC(unsigned long long)
 
-inline __device__       unsigned int device_atomic_fetch_inc_mod(  unsigned int* ptr,       unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) { return atomicInc(ptr,  val); }
-inline __device__       unsigned int device_atomic_fetch_dec_mod(  unsigned int* ptr,       unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) { return atomicDec(ptr,  val); }
-// clang-format on
+#undef DESUL_IMPL_HIP_ATOMIC_FETCH_INC
 
-#define DESUL_IMPL_HIP_DEVICE_ATOMIC_FETCH_OP(OP, TYPE)                                \
+#define DESUL_IMPL_HIP_ATOMIC_FETCH_INC_MOD(MEMORY_SCOPE, MEMORY_SCOPE_STRING_LITERAL) \
   template <class MemoryOrder>                                                         \
-  __device__ TYPE device_atomic_fetch_##OP(                                            \
-      TYPE* ptr, TYPE val, MemoryOrder, MemoryScopeDevice) {                           \
-    __threadfence();                                                                   \
-    TYPE return_val =                                                                  \
-        device_atomic_fetch_##OP(ptr, val, MemoryOrderRelaxed(), MemoryScopeDevice()); \
-    __threadfence();                                                                   \
-    return return_val;                                                                 \
+  __device__ inline unsigned int device_atomic_fetch_inc_mod(                          \
+      unsigned int* ptr, unsigned int val, MemoryOrder, MEMORY_SCOPE) {                \
+    return __builtin_amdgcn_atomic_inc32(                                              \
```

Comment:
I was wondering whether we should worry about the use of builtins that are tied to the GCN microarchitectures.
I looked in the [doc for the LLVM backend for AMD GPUs](https://llvm.org/docs/AMDGPUUsage.html#target-triples) and it looks like all GPUs released in the last decade(?) are based on successive generations of that microarchitecture.  The doc says it is for "AMD GPUs GCN GFX6 onwards for graphics and compute shaders".